drivers/led/is31fl3194/is31fl3194_reg.c

```diff
@@ -0,0 +1,837 @@
+
+static const struct is31fl3194_px_t2_t3 px_t2_t3_default = { .t2 = 0b0000, .t3 = 0b0000 };
+
+static const struct is31fl3194_px_tp_t4 px_tp_t4_default = { .tp = 0b0000, .t4 = 0b0000 };
+
+static const struct is31fl3194_px_col_en px_col_en_default = { .ce1 = 0b1, .ce2 = 0b0, .ce3 = 0b0 };
+
+static const struct is31fl3194_px_ccty px_ccty_default = { .cct1 = 0b00,
+							   .cct2 = 0b00,
+							   .cct3 = 0b00 };
+
+static const struct is31fl3194_px_nxt px_nxt_default = { .nxtx = 0b01,
+							 .gamx = 0b00,
+							 .mtpltx = 0b0000 };
+
+static const struct is31fl3194_px_lt px_lt_default = { .pltx_l = 0b0000000, .pltx_h = 0b0 };
+
+static const struct is31fl3194_regs regs_default = {
+	.ops_conf = ops_conf_default,
+	.out_en = out_en_default,
+	.cbx = cbx_default,
+	.hfx = hfx_default,
+	.outx_cl = { outx_cl_default },
+	.px_coly_oz = { { { px_coly_oz_default } } },
+	.px_ts_t1 = { px_ts_t1_default },
+	.px_t2_t3 = { px_t2_t3_default },
+	.px_tp_t4 = { px_tp_t4_default },
+	.px_col_en = { px_col_en_default },
+	.px_ccty = { px_ccty_default },
+	.px_nxt = { px_nxt_default },
+	.px_lt = { px_lt_default },
+};
+
+static uint8_t px_state_reg_addr[] = { IS31FL3194_P1_STATE, IS31FL3194_P2_STATE,
+				       IS31FL3194_P3_STATE };
+
+static uint8_t outx_cl_reg_addr[] = { IS31FL3194_OUT1_CURRENT_LEVEL, IS31FL3194_OUT2_CURRENT_LEVEL,
+				      IS31FL3194_OUT3_CURRENT_LEVEL };
+
+static uint8_t colx_py_base_reg_addr[3][3] = {
+	{ IS31FL3194_COLOR_1_P1_BASE, IS31FL3194_COLOR_1_P2_BASE, IS31FL3194_COLOR_1_P3_BASE },
+	{ IS31FL3194_COLOR_2_P1_BASE, IS31FL3194_COLOR_2_P2_BASE, IS31FL3194_COLOR_2_P3_BASE },
+	{ IS31FL3194_COLOR_3_P1_BASE, IS31FL3194_COLOR_3_P2_BASE, IS31FL3194_COLOR_3_P3_BASE },
+};
+
+static uint8_t px_ts_and_t1_reg_addr[] = { IS31FL3194_P1_TS_AND_T1, IS31FL3194_P2_TS_AND_T1,
+					   IS31FL3194_P3_TS_AND_T1 };
+
+static uint8_t px_t2_and_t3_reg_addr[] = { IS31FL3194_P1_T2_AND_T3, IS31FL3194_P2_T2_AND_T3,
+					   IS31FL3194_P3_T2_AND_T3 };
+
+static uint8_t px_tp_and_t4_reg_addr[] = { IS31FL3194_P1_TP_AND_T4, IS31FL3194_P2_TP_AND_T4,
+					   IS31FL3194_P3_TP_AND_T4 };
+
+const uint8_t px_color_en_reg_addr[] = {
+	IS31FL3194_P1_COLOR_EN,
+	IS31FL3194_P2_COLOR_EN,
+	IS31FL3194_P3_COLOR_EN,
+};
+
+static uint8_t px_color_cycle_times_reg_addr[] = {
+	IS31FL3194_P1_COLOR_CYCLE_TIMES,
+	IS31FL3194_P2_COLOR_CYCLE_TIMES,
+	IS31FL3194_P3_COLOR_CYCLE_TIMES,
+};
+
+static uint8_t px_nxt_reg_addr[] = {
+	IS31FL3194_P1_NXT,
+	IS31FL3194_P2_NXT,
+	IS31FL3194_P3_NXT,
+};
+
+static uint8_t px_loop_times_reg_addr[] = {
+	IS31FL3194_P1_LOOP_TIMES,
+	IS31FL3194_P2_LOOP_TIMES,
+	IS31FL3194_P3_LOOP_TIMES,
+};
+
+static uint8_t px_update_reg_addr[] = { IS31FL3194_P1_UPDATE, IS31FL3194_P2_UPDATE,
+					IS31FL3194_P3_UPDATE };
+
+static int read_reg(const struct is31fl3194_ctx *ctx, uint8_t reg, uint8_t *data, uint16_t len)
+{
+	return ctx->read_reg(ctx->handle, reg, data, len);
```

Comment:
As I said in another comment, I don't see the purpose for this indirection.

Reply:
NACK. These kind of HAL abstractions makes sense if you are creating a library that might be used by different drivers. But here, there is obviously one driver making use of it and the HAL indirection is really making the code look complex. Please get rid of it.